### ISSUE TITLE
[css-grid] Serialization of grid-template-rows/columns

### ISSUE THREAD

**Issue by ewilligers** [w3c/csswg-drafts#4173]
If we set `grid-template-rows` to `"repeat(1, [] 10px)"`, and then read the specified value, what should the result be? (All browsers give `"10px"` as the computed value.)

* `"repeat(1, [] 10px)"` - Blink
* `"repeat(1, 10px)"` - Firefox
* `"[] 10px"` - Safari
* `"10px"` Edge 18

Edge 18 only "expands" `repeat(1, ...)`, while Safari expands any `repeat(n, ...)` in the specified value. (Blink, Firefox and Safari expand `repeat(n, ...)` in computed values.)

**Comment by MatsPalmgren** [comment]
> (All browsers give "10px" as the computed value.)

FYI, @emilio [fixed](https://bugzilla.mozilla.org/show_bug.cgi?id=1519958) Gecko yesterday to preserve `repeat` in the computed value.

**Comment by Loirooriol** [comment]
For the specified value, Blink used to expand integral `repeat()` until https://crbug.com/716114. I also fixed WebKit in https://webkit.org/b/197840. So both should output `"repeat(1, [] 10px)"`.

Regarding `[]`, I don't know whether it should be preserved in the specified value. But I don't think it should in the resolved value, because from https://drafts.csswg.org/css-grid/#computed-tracks

> A computed track list is a list alternating between line name sets and track sections, with the first and last items being line name sets.

> A line name set is a (potentially empty) set of identifiers representing line names.

This means that if you specify `1px [] 2px 3px`, the computed value will be `[] 1px [] 2px [] 3px []`. Including the extra `[]` in the serialization would be unexpected IMO, so skipping them all seems better.

Getting back to `repeat()`, Blink and WebKit still expand integral repeat() in the computed value. This is wrong according to the spec. However, note that for grid containers, the resolved value is the used one, with `repeat()` expanded.

I'm not a fan of special casing `repeat(1`.

To summarize, I think the expected serializations should be:
 - Specified value: `"repeat(1, [] 10px)"` or `"repeat(1, 10px)"`, dunno
 - Computed value (resolved value for non-grids): `"repeat(1, 10px)"`
 - Used value (resolved value for grids): `"10px"`

**Comment by ewilligers** [comment]
> Including the extra `[]` in the serialization would be unexpected IMO, so skipping them all seems better.

OK, I'll update Blink and WPTs to drop `[]` in serialization.

I'll add computed-001.html and computed-002.html for non-grid and grid gCS expectations.

**Comment by Loirooriol** [comment]
> OK, I'll update Blink and WPTs to drop `[]` in serialization.

Blink already drops it in the resolved value, https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/css/properties/computed_style_utils.cc?rcl=7aba141f9619adf5e4efe8b09b9c453f451754e3&l=1229-1230. Will you remove it in specified values?

> I'll add computed-001.html and computed-002.html for non-grid and grid gCS expectations.

In https://chromium-review.googlesource.com/c/chromium/src/+/1724517 I have some tests of this kind (not with integral repeats, I planned to add them later, nor `[]`). You can copy that if you want the tests to be in wpt before my patch is ready to land.

**Comment by ewilligers** [comment]
Consensus is
 - Specified value: `"repeat(1, 10px)"`
 - Computed value (resolved value for non-grids): `"repeat(1, 10px)"`
 - Used value (resolved value for grids): `"10px"`
as implemented in Firefox.
